### PR TITLE
[BEAM-10317] Python - Update BigQueryIO to tag BigQuery Jobs with the Dataflow Job ID

### PR DESCRIPTION
**Pull request by ajamato**
[BEAM-10317] Python - Update BigQueryIO to tag BigQuery Jobs with the Dataflow Job ID

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark
--- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ajamato** [comment]
@pabloem @chamikaramj

**Comment by pabloem** [comment]
can you address formatter / lint / rat / test  errors?

**Comment by ajamato** [comment]
retest this please

**Comment by ajamato** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by aaltay** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by ajamato** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by pabloem** [comment]
Run Python 3.5 PostCommit

File: sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  """Metadata class for BigQueryIO. i.e. to use as BQ job labels.
+
+  Do not construct directly, use the CreateBigQueryIOMetadata factory.
+  Which will request metadata properly based on which runner is being used.
+  """
+  def __init__(self, beam_job_id=None):
+    self.beam_job_id = beam_job_id
+
+  def add_additional_bq_job_labels(self, job_labels=None):
+    job_labels = dict() if job_labels is None else job_labels
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py:52]: (nit)Up to you, but you could reduce to this:

```suggestion
    job_labels = job_labels or {}
```

File: sdks/python/apache_beam/io/gcp/bigquery_file_loads.py [hunk 2]
```diff
@@ -327,10 +329,13 @@ def __init__(
     self.write_disposition = write_disposition
     self.test_client = test_client
     self._observed_tables = set()
+    self.bq_io_metadata = None
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery_file_loads.py:332]: Do you want to add the display data item for this DoFn?

File: sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+"""Metadata for use in BigQueryIO, i.e. a job_id to use in BQ job labels."""
+
+# pytype: skip-file
+
+from __future__ import absolute_import
+
+from apache_beam.io.gcp import gce_metadata_util
+
+
+def CreateBigQueryIOMetadata():
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py:27]: we usually make functions be snake_case. Could you change this please?

File: sdks/python/apache_beam/io/gcp/gce_metadata_util.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+from __future__ import absolute_import
+
+import httplib2
+
+from apache_beam.internal.http_client import get_new_http
+
+BASE_METADATA_URL = "http://metadata/computeMetadata/v1/"
+
+
+def FetchMetadata(key):
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/gce_metadata_util.py:31]: make functions snake_case please

File: sdks/python/apache_beam/io/gcp/gce_metadata_util.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  try:
+    h = get_new_http(timeout_secs=5)
+    headers = {"Metadata-Flavor": "Google"}
+    uri = BASE_METADATA_URL + key
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/gce_metadata_util.py:35]: `os.path.join` should ensure URLs are joined properly independently of whether the key has/does not have a leading slash. Do you think that'd be good to add?

File: sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  Do not construct directly, use the CreateBigQueryIOMetadata factory.
+  Which will request metadata properly based on which runner is being used.
+  """
+  def __init__(self, beam_job_id=None):
+    self.beam_job_id = beam_job_id
+
+  def add_additional_bq_job_labels(self, job_labels=None):
+    job_labels = dict() if job_labels is None else job_labels
+    if self.beam_job_id and 'beam_job_id' not in job_labels:
+      job_labels['beam_job_id'] = self.beam_job_id
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/bigquery_io_metadata.py:54]: this means that for other runners, BQ will receive jobs with a label of `'beam_job_id'=None` - is that fine?

File: sdks/python/apache_beam/io/gcp/gce_metadata_util.py (hunk 5 above)
Comment (pabloem) [reply sdks/python/apache_beam/io/gcp/gce_metadata_util.py:35]: (up to you)

File: sdks/python/apache_beam/io/gcp/gce_metadata_util.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    resp, content = h.request(uri, "GET", headers=headers)
+    if resp.status == 200:
+      return content
+  except httplib2.HttpLib2Error:
```
Comment (pabloem) [inline sdks/python/apache_beam/io/gcp/gce_metadata_util.py:39]: how will it work for non-dataflow workers? Will it fail immediately? time out?

**Comment by pabloem** [comment]
Run Python 3.5 PostCommit

File: sdks/python/apache_beam/io/gcp/gce_metadata_util.py (hunk 7 above)
Comment (ajamato) [reply sdks/python/apache_beam/io/gcp/gce_metadata_util.py:39]: The URL should not resolve, so it should fail quickly after attempting a request, not for the full timeout. Ideally we are making this request as infrequently as possible.

I've also just added a method to validate the label, in the off case that the metadata server is giving bad output. I don't want this to make the BQ jobs fail (they will fail if given invalid labels).

**Comment by pabloem** [comment]
this LGTM - please address lint issues, and I think it's good to go - https://scans.gradle.com/s/enabhz4kd26o2/console-log?task=:sdks:python:test-suites:tox:py37:lintPy37

**Comment by pabloem** [comment]
Run Python 3.5 PostCommit

**Comment by ajamato** [comment]
Thanks Pablo. I addressed the lint issues and pushed the changes. Hopefully it passes

**Comment by pabloem** [comment]
Run Python 3.5 PostCommit

**Comment by pabloem** [comment]
Run Python 2 PostCommit

**Comment by pabloem** [comment]
Thanks @ajamato - LGTM

**Pull request merged**